File: backend/app/api/v1/endpoints/offers.py
"""API endpoints for offers marketplace."""

import uuid
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_db, get_current_user, get_current_director
from app.models.user import User, UserRole
from app.models.offer import Offer
from app.models.purchase import Purchase, PurchaseStatus
from app.models.student import StudentProfile
from app.schemas.offer import OfferCreate, OfferRead, OfferUpdate, OfferListRead
from app.schemas.purchase import PurchaseRead, StudentBalanceRead

router = APIRouter()


@router.get("", response_model=List[OfferListRead])
async def get_offers(
    category: Optional[str] = Query(None, description="Filter by category"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get all active offers."""
    query = db.query(Offer).filter(Offer.is_active == True)  # noqa: E712

    if category:
        query = query.filter(Offer.category == category)

    offers = query.order_by(Offer.created_at.desc()).all()

    return [
        OfferListRead(
            id=o.id,
            name=o.name,
            description=o.description,
            price=o.price,
            image_url=o.image_url,
            brand_name=o.brand_name,
            category=o.category,
            stock_quantity=o.stock_quantity,
            is_available=o.is_active
            and (o.stock_quantity is None or o.stock_quantity > 0),
        )
        for o in offers
    ]


@router.get("/balance", response_model=StudentBalanceRead)
async def get_my_balance(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get current student's bonus points balance."""
    if current_user.role != UserRole.STUDENT:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only students can view balance",
        )

    student = (
        db.query(StudentProfile)
        .filter(StudentProfile.user_id == current_user.id)
        .first()
    )

    if not student:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Student profile not found",
        )

    return StudentBalanceRead(bonus_points=student.bonus_points or 0)


@router.get("/{offer_id}", response_model=OfferRead)
async def get_offer(
    offer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get offer details."""
    offer = db.query(Offer).filter(Offer.id == offer_id).first()
    if not offer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Offer not found",
        )
    return offer


@router.post("", response_model=OfferRead, status_code=status.HTTP_201_CREATED)
async def create_offer(
    offer_in: OfferCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_director),
):
    """Create a new offer (Director/Admin only)."""
    offer = Offer(**offer_in.model_dump())
    db.add(offer)
    db.commit()
    db.refresh(offer)
    return offer


@router.put("/{offer_id}", response_model=OfferRead)
async def update_offer(
    offer_id: int,
    offer_in: OfferUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_director),
):
    """Update an offer (Director/Admin only)."""
    offer = db.query(Offer).filter(Offer.id == offer_id).first()
    if not offer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Offer not found",
        )

    update_data = offer_in.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(offer, field, value)

    db.commit()
    db.refresh(offer)
    return offer


@router.delete("/{offer_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_offer(
    offer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_director),
):
    """Soft delete an offer (Director/Admin only)."""
    offer = db.query(Offer).filter(Offer.id == offer_id).first()
    if not offer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Offer not found",
        )

    offer.is_active = False
    db.commit()
    return None


@router.post("/{offer_id}/purchase", response_model=PurchaseRead)
async def purchase_offer(
    offer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Purchase an offer (Student only) - atomic transaction."""
    if current_user.role != UserRole.STUDENT:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only students can make purchases",
        )

    # Get student profile with row lock for update
    student = (
        db.query(StudentProfile)
        .filter(StudentProfile.user_id == current_user.id)
        .with_for_update()
        .first()
    )

    if not student:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Student profile not found",
        )

    # Get offer with row lock
    offer = (
        db.query(Offer)
        .filter(Offer.id == offer_id, Offer.is_active == True)  # noqa: E712
        .with_for_update()
        .first()
    )

    if not offer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Offer not found or inactive",
        )

    # Check stock
    if offer.stock_quantity is not None and offer.stock_quantity <= 0:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Offer out of stock",
        )

    # Check balance
    current_balance = student.bonus_points or 0
    if current_balance < offer.price:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Insufficient balance. Need {offer.price}, have {current_balance}",
        )

    # Generate unique QR code
    qr_code = f"IMUSUM-{uuid.uuid4().hex[:12].upper()}"

    # Atomic transaction
    try:
        # Deduct points
        student.bonus_points = current_balance - offer.price

        # Reduce stock if applicable
        if offer.stock_quantity is not None:
            offer.stock_quantity -= 1

        # Create purchase record
        purchase = Purchase(
            student_id=student.id,
            offer_id=offer.id,
            points_spent=offer.price,
            qr_code=qr_code,
            status=PurchaseStatus.PENDING.value,
        )
        db.add(purchase)
        db.commit()
        db.refresh(purchase)

        return PurchaseRead(
            id=purchase.id,
            offer_id=purchase.offer_id,
            points_spent=purchase.points_spent,
            qr_code=purchase.qr_code,
            status=purchase.status,
            created_at=purchase.created_at,
            redeemed_at=purchase.redeemed_at,
            offer_name=offer.name,
            offer_brand=offer.brand_name,
            offer_image_url=offer.image_url,
        )
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Purchase failed. Please try again.",
        )

File: backend/src/routes/v1/offers.ts
import crypto from 'crypto';

import type { FastifyPluginAsync } from 'fastify';

import { DIRECTOR_PLUS_ROLES } from '../../lib/auth.js';
import { badRequest, forbidden, internalServerError, notFound } from '../../lib/errors.js';
import { prisma } from '../../lib/prisma.js';
import { serializeOffer, serializeOfferListItem, serializePurchase } from '../../lib/serializers.js';

const offersRoutes: FastifyPluginAsync = async (fastify) => {
  fastify.get('', { preHandler: [fastify.authenticate] }, async (request) => {
    const query = request.query as { category?: string };

    const offers = await prisma.offer.findMany({
      where: {
        is_active: true,
        ...(query.category ? { category: query.category } : {}),
      },
      orderBy: { created_at: 'desc' },
    });

    return offers.map(serializeOfferListItem);
  });

  fastify.get('/balance', { preHandler: [fastify.authenticate] }, async (request) => {
    const currentUser = request.currentUser!;
    if (currentUser.role !== 'student') {
      forbidden('Only students can view balance');
    }

    const student = await prisma.studentProfile.findUnique({ where: { user_id: currentUser.id } });
    if (!student) {
      notFound('Student profile not found');
    }

    return { bonus_points: student.bonus_points ?? 0 };
  });

  fastify.get('/:offer_id', { preHandler: [fastify.authenticate] }, async (request) => {
    const params = request.params as { offer_id: string };
    const offerId = Number(params.offer_id);

    const offer = await prisma.offer.findUnique({ where: { id: offerId } });
    if (!offer) {
      notFound('Offer not found');
    }

    return serializeOffer(offer);
  });

  fastify.post('', { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] }, async (request, reply) => {
    const body = request.body as {
      name: string;
      description?: string;
      price: number;
      image_url?: string;
      brand_name: string;
      category?: string;
      stock_quantity?: number;
    };

    const offer = await prisma.offer.create({
      data: {
        name: body.name,
        description: body.description ?? null,
        price: body.price,
        image_url: body.image_url ?? null,
        brand_name: body.brand_name,
        category: body.category ?? 'other',
        stock_quantity: body.stock_quantity ?? null,
        created_at: new Date(),
      },
    });

    return reply.status(201).send(serializeOffer(offer));
  });

  fastify.put(
    '/:offer_id',
    { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] },
    async (request) => {
      const params = request.params as { offer_id: string };
      const body = request.body as {
        name?: string;
        description?: string;
        price?: number;
        image_url?: string;
        brand_name?: string;
        category?: string;
        stock_quantity?: number | null;
        is_active?: boolean;
      };

      const offerId = Number(params.offer_id);
      const existing = await prisma.offer.findUnique({ where: { id: offerId } });
      if (!existing) {
        notFound('Offer not found');
      }

      const offer = await prisma.offer.update({
        where: { id: offerId },
        data: {
          ...(body.name !== undefined ? { name: body.name } : {}),
          ...(body.description !== undefined ? { description: body.description } : {}),
          ...(body.price !== undefined ? { price: body.price } : {}),
          ...(body.image_url !== undefined ? { image_url: body.image_url } : {}),
          ...(body.brand_name !== undefined ? { brand_name: body.brand_name } : {}),
          ...(body.category !== undefined ? { category: body.category } : {}),
          ...(body.stock_quantity !== undefined ? { stock_quantity: body.stock_quantity } : {}),
          ...(body.is_active !== undefined ? { is_active: body.is_active } : {}),
        },
      });

      return serializeOffer(offer);
    }
  );

  fastify.delete(
    '/:offer_id',
    { preHandler: [fastify.requireRoles(DIRECTOR_PLUS_ROLES, 'Director access required')] },
    async (request, reply) => {
      const params = request.params as { offer_id: string };
      const offerId = Number(params.offer_id);

      const existing = await prisma.offer.findUnique({ where: { id: offerId } });
      if (!existing) {
        notFound('Offer not found');
      }

      await prisma.offer.update({ where: { id: offerId }, data: { is_active: false } });
      return reply.status(204).send();
    }
  );

  fastify.post('/:offer_id/purchase', { preHandler: [fastify.authenticate] }, async (request) => {
    const params = request.params as { offer_id: string };
    const offerId = Number(params.offer_id);
    const currentUser = request.currentUser!;

    if (currentUser.role !== 'student') {
      forbidden('Only students can make purchases');
    }

    try {
      const result = await prisma.$transaction(async (tx) => {
        const student = await tx.studentProfile.findUnique({ where: { user_id: currentUser.id } });
        if (!student) {
          notFound('Student profile not found');
        }

        await tx.$queryRaw`SELECT id FROM student_profiles WHERE id = ${student.id} FOR UPDATE`;

        const offer = await tx.offer.findFirst({ where: { id: offerId, is_active: true } });
        if (!offer) {
          notFound('Offer not found or inactive');
        }

        await tx.$queryRaw`SELECT id FROM offers WHERE id = ${offer.id} FOR UPDATE`;

        if (offer.stock_quantity !== null && offer.stock_quantity <= 0) {
          badRequest('Offer out of stock');
        }

        const currentBalance = student.bonus_points ?? 0;
        if (currentBalance < offer.price) {
          badRequest(`Insufficient balance. Need ${offer.price}, have ${currentBalance}`);
        }

        const qrCode = `IMUSUM-${crypto.randomUUID().replace(/-/g, '').slice(0, 12).toUpperCase()}`;

        await tx.studentProfile.update({
          where: { id: student.id },
          data: { bonus_points: currentBalance - offer.price },
        });

        if (offer.stock_quantity !== null) {
          await tx.offer.update({
            where: { id: offer.id },
            data: { stock_quantity: offer.stock_quantity - 1 },
          });
        }

        const purchase = await tx.purchase.create({
          data: {
            student_id: student.id,
            offer_id: offer.id,
            points_spent: offer.price,
            qr_code: qrCode,
            status: 'pending',
            created_at: new Date(),
          },
        });

        return { purchase, offer };
      });

      return serializePurchase(result.purchase, {
        name: result.offer.name,
        brand_name: result.offer.brand_name,
        image_url: result.offer.image_url,
      });
    } catch (error) {
      if (error && typeof error === 'object' && 'statusCode' in error) {
        throw error;
      }
      request.log.error({ err: error }, 'Purchase failed');
      internalServerError('Purchase failed. Please try again.');
    }
  });
};

export default offersRoutes;
